src/tokens/stream_reader.rs

```diff
@@ -0,0 +1,539 @@
+                            // container values are not stepped into
+                            StreamItem::Value(_) => Some(Instruction::NextEnd),
+                            StreamItem::Null(_) => {
+                                unreachable!("Cannot be null and on a start container")
+                            }
+                            // if our current item is nothing, we've stepped in
+                            StreamItem::Nothing => Some(Instruction::Next),
+                        }
+                    }
+                    // end positions are terminal unless step_out if applicable
+                    Content::EndContainer(_) | Content::EndStream => None,
+                };
+                drop(token);
+                // we need to drop the cell borrow before we can advance
+                match insn_opt {
+                    Some(instruction) => self.next_token(instruction),
+                    None => Ok(StreamItem::Nothing),
+                }
+            }
+        }
+    }
+
+    fn current(&self) -> Self::Item {
+        self.curr_item
+    }
+
+    fn ion_type(&self) -> Option<IonType> {
+        match (self.curr_item, &self.curr_token_cell) {
+            (StreamItem::Nothing, _) => None,
+            (_, None) => None,
+            (_, Some(token_cell)) => {
+                let token = token_cell.borrow();
+                match token.content() {
+                    Content::Null(ion_type) => Some(*ion_type),
+                    Content::Scalar(thunk) => Some(thunk.scalar_type().into()),
+                    Content::StartContainer(container_type) => Some((*container_type).into()),
+                    Content::EndContainer(_) => {
+                        // the end of a container is considered not positioned
+                        None
+                    }
+                    Content::EndStream => None,
+                }
+            }
+        }
+    }
+
+    fn annotations<'b>(&'b self) -> Box<dyn Iterator<Item = IonResult<Self::Symbol>> + 'b> {
+        let iter = match &self.curr_token_cell {
+            None => Annotations::empty().into_iter(),
+            Some(token_cell) => {
+                let mut token = token_cell.borrow_mut();
+                let annotations_result = token.annotations.no_memoize();
+                match annotations_result {
+                    Ok(annotations) => annotations.into_iter(),
+                    Err(e) => {
+                        // if this happens we return a singleton error
+                        return Box::new([Err(e)].into_iter());
+                    }
+                }
+            }
+        };
+        Box::new(iter.map(Ok))
+    }
+
+    fn field_name(&self) -> IonResult<Self::Symbol> {
+        match &self.curr_token_cell {
+            None => illegal_operation(NO_FIELD_NAME_ERROR_TEXT),
+            Some(token_cell) => {
+                let mut token = token_cell.borrow_mut();
+                match token.share_field_name() {
```

Comment:
🗺️ We could be more aggressive about _not_ interning, only intern when this is called after we materialized it once, but I don't think that is worth it, especially while we're re-thinking the `IonReader` interface in #536.